src/blocks/mha.py

```diff
@@ -0,0 +1,72 @@
+import torch
+import torch.nn as nn
+from einops import rearrange
+from torch import Tensor
+
+from src.interfaces import AttentionBase
+
+
+class MultiHeadAttention(AttentionBase):
+    """Multi-head attention with optional masking."""
+
+    def __init__(
+        self,
+        dim: int,
+        n_heads: int,
+        embed_bias: bool = True,
+        q_dim: int = None,
+        ctx_dim: int = None,
+        dropout: float = 0.0,
+    ):
+        """Initialize multi-head attention.
+
+        Args:
+            dim: Token dimensionality.
+            n_heads: Number of attention heads.
+            embed_bias: Whether to use bias in linear projections.
+            q_dim: Query dimensionality (defaults to dim).
+            ctx_dim: Context dimensionality (defaults to dim).
+            dropout: Dropout rate.
+        """
+        super().__init__()
+
+        assert dim % n_heads == 0, "dim must be divisible by n_heads"
+
+        self.n_heads = n_heads
+        q_dim = q_dim or dim
+        ctx_dim = ctx_dim or dim
+        dim_head = dim // n_heads
+        self.scale = dim_head**-0.5
+
+        self.query = torch.nn.Linear(q_dim, dim, bias=embed_bias)
+        self.key = torch.nn.Linear(ctx_dim, dim, bias=embed_bias)
+        self.value = torch.nn.Linear(ctx_dim, dim, bias=embed_bias)
+        self.output = torch.nn.Linear(dim, q_dim, bias=embed_bias)
+
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(
+        self,
+        q: Tensor,
+        ctx: Tensor,
+        mask: torch.Tensor = None,
+    ) -> tuple[Tensor, Tensor]:
+        q, k, v = map(lambda f, x: f(x), (self.query, self.key, self.value), (q, ctx, ctx))
+        q, k, v = map(lambda x: rearrange(x, "b s (h d) -> b h s d", h=self.n_heads), (q, k, v))
+
+        scores = torch.einsum("b h s d, b h t d -> b h s t", q, k) * self.scale
```

Comment:
With each `einsum` expression, I would consider writing a short comment what it does, for people who are not so familiar with `einsum`. Although this here should be straight forward :)


```suggestion
        # Q @ K.T
        scores = torch.einsum("b h s d, b h t d -> b h s t", q, k) * self.scale
```